grpc/controlproxy/controlproxy.go

```diff
@@ -0,0 +1,245 @@
+	}
+	if reply.Error != nil {
+		return nil, common.ToSnapError(reply.Error)
+	}
+	nss := ToNSS(reply.NSS)
+	return nss, nil
+}
+func (c ControlProxy) PublishMetrics(contentType string, content []byte, pluginName string, pluginVersion int, config map[string]ctypes.ConfigValue, taskID string) []error {
+	req := GetPubProcReq(contentType, content, pluginName, pluginVersion, config, taskID)
+	reply, err := c.Client.PublishMetrics(context.Background(), req)
+	errs := make([]error, 0)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	rerrs := ReplyErrorsToErrors(reply.Errors)
+	errs = append(errs, rerrs...)
+	return errs
+}
+
+func (c ControlProxy) ProcessMetrics(contentType string, content []byte, pluginName string, pluginVersion int, config map[string]ctypes.ConfigValue, taskID string) (string, []byte, []error) {
+	req := GetPubProcReq(contentType, content, pluginName, pluginVersion, config, taskID)
+	reply, err := c.Client.ProcessMetrics(context.Background(), req)
+	errs := make([]error, 0)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	rerrs := ReplyErrorsToErrors(reply.Errors)
+	errs = append(errs, rerrs...)
+	return reply.ContentType, reply.Content, errs
+}
+
+func (c ControlProxy) CollectMetrics(mts []core.Metric, deadline time.Time, taskID string) ([]core.Metric, []error) {
+	req := &rpc.CollectMetricsRequest{
+		Metrics: common.NewMetrics(mts),
+		Deadline: &common.Time{
+			Sec:  deadline.Unix(),
+			Nsec: int64(deadline.Nanosecond()),
+		},
+		TaskID: taskID,
+	}
+	reply, err := c.Client.CollectMetrics(context.Background(), req)
+	errs := make([]error, 0)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+	rerrs := ReplyErrorsToErrors(reply.Errors)
+	if len(rerrs) > 0 {
+		errs = append(errs, rerrs...)
+		return nil, errs
+	}
+	metrics := common.CommonMetricsToCore(reply.Metrics)
+	return metrics, nil
+}
+
+func (c ControlProxy) GetPluginContentTypes(n string, t core.PluginType, v int) ([]string, []string, error) {
+	req := &rpc.GetPluginContentTypesRequest{
+		Name:       n,
+		PluginType: getPluginType(t),
+		Version:    int32(v),
+	}
+	reply, err := c.Client.GetPluginContentTypes(context.Background(), req)
+	if err != nil {
+		return nil, nil, err
+	}
+	if reply.Error != "" {
+		return nil, nil, errors.New(reply.Error)
+	}
+	return reply.AcceptedTypes, reply.ReturnedTypes, nil
+}
+
+func (c ControlProxy) ValidateDeps(mts []core.Metric, plugins []core.SubscribedPlugin) []serror.SnapError {
+	req := &rpc.ValidateDepsRequest{
+		Metrics: common.NewMetrics(mts),
+		Plugins: common.SubPluginsToMsg(plugins),
+	}
+	reply, err := c.Client.ValidateDeps(context.Background(), req)
+	serrs := groupErrors(err, reply.Errors)
+	return serrs
+}
+
+func (c ControlProxy) SubscribeDeps(taskID string, mts []core.Metric, plugins []core.Plugin) []serror.SnapError {
+	req := depsRequest(taskID, mts, plugins)
+	reply, err := c.Client.SubscribeDeps(context.Background(), req)
+	serrs := groupErrors(err, reply.Errors)
```

Comment:
Need to ensure that reply is not nil before referencing it. This goes for all the request/replies in this file.